### PR TITLE
Redirect isSteppingCarefully for better sponge compatibility

### LINKED ISSUE #135: Compatibility Issue with Sponge
https://github.com/SpongePowered/Sponge/issues/3514

There appears to be a code conflicts between your mod and Sponge... Is it possible this can be fixed?

### PR DESCRIPTION
**Pull request by gabizou**
Should fix #135 

As suggested in https://github.com/SpongePowered/Sponge/issues/3514#issuecomment-1053762787

### CODE REVIEW

File: src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java [hunk 1]
```diff
@@ -131,10 +133,11 @@ public void read(CompoundNBT compound, CallbackInfo info) {
...
-	@Redirect(method = "move", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/Block;stepOn(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/entity/Entity;)V"))
-	private void onEntityWalk(Block block, World world, BlockPos pos, Entity entity) {
-		if (!EntityWalkEvent.onEntityWalk(world, pos, entity)) {
-			block.stepOn(world, pos, entity);
+	@Redirect(method = "move", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/Entity;isSteppingCarefully()Z"))
+	private boolean checkIsSteppingCarefullyAndCanEntityWalkOn(Entity instance) {
+		if (!instance.isSteppingCarefully()) {
+			return false;
 		}
+		return !EntityWalkEvent.onEntityWalk(this.level, this.getOnPos(), instance);
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java]:
```suggestion
		return instance.isSteppingCarefully() && !EntityWalkEvent.onEntityWalk(this.level, this.getOnPos(), instance);
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java [hunk 2]
```diff
@@ -131,10 +133,11 @@ public void read(CompoundNBT compound, CallbackInfo info) {
 		}
 	}
 
-	@Redirect(method = "move", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/Block;stepOn(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/entity/Entity;)V"))
-	private void onEntityWalk(Block block, World world, BlockPos pos, Entity entity) {
-		if (!EntityWalkEvent.onEntityWalk(world, pos, entity)) {
-			block.stepOn(world, pos, entity);
+	@Redirect(method = "move", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/Entity;isSteppingCarefully()Z"))
+	private boolean checkIsSteppingCarefullyAndCanEntityWalkOn(Entity instance) {
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java]: This is a pretty verbose name.
```suggestion
	private boolean onIsSteppingCarefully(Entity instance) {
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java [hunk 3]
```diff
@@ -28,9 +27,12 @@
 import java.util.Set;
 
 @Mixin(Entity.class)
-public final class EntityMixin implements IDataManager {
+public abstract class EntityMixin implements IDataManager {
 	@Shadow
-	private World level;
+	public World level;
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java:32]: Keep this private.
```suggestion
	private World level;
```

**Review by SmellyModder — CHANGES_REQUESTED** [review]

**Comment by SmellyModder** [comment]
Thanks for the fix. I'll also port the fix to 1.18.2 when we finish the port.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java (hunk 3 above)
Comment (gabizou) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/EntityMixin.java:32]: but the field itself is public, Shadows should normally retain the same access level as their targets, regardless whether their intended uses are internal or external to the mixin in question.

**Comment by MrJorman** [comment]
pls give me 1.16.5 .jar pls

**Review by SmellyModder — APPROVED** [review]

**Comment by thebikwirm** [comment]
has this been fixed
in 16.5 within recent updates?

**Pull request merged**